airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/AsyncStreamConsumer.java

```diff
@@ -121,55 +118,32 @@ public void accept(final String messageString, final Integer sizeInBytes) throws
    * Deserializes to a {@link PartialAirbyteMessage} which can represent both a Record or a State
    * Message
    *
+   * PartialAirbyteMessage holds either:
+   * <li>entire serialized message string when message is a valid State Message
+   * <li>serialized AirbyteRecordMessage when message is a valid Record Message</li>
+   *
    * @param messageString the string to deserialize
    * @return PartialAirbyteMessage if the message is valid, empty otherwise
    */
-  private Optional<PartialAirbyteMessage> deserializeAirbyteMessage(final String messageString) {
+  @VisibleForTesting
+  public static Optional<PartialAirbyteMessage> deserializeAirbyteMessage(final String messageString) {
+    // TODO: (ryankfu) plumb in the serialized AirbyteStateMessage to match AirbyteRecordMessage code
+    // parity
     final Optional<PartialAirbyteMessage> messageOptional = Jsons.tryDeserialize(messageString, PartialAirbyteMessage.class)
-        .map(partial -> partial.withSerialized(messageString));
+        .map(partial -> {
+          if (partial.getType().equals(Type.RECORD) && partial.getRecord().getData() != null) {
+            return partial.withSerialized(partial.getRecord().getData().toString());
+          } else if (partial.getType().equals(Type.STATE)) {
+            return partial.withSerialized(messageString);
+          } else {
+            return null;
+          }
+        });
+
     if (messageOptional.isPresent()) {
       return messageOptional;
-    } else {
-      if (isStateMessage(messageString)) {
-        throw new IllegalStateException("Invalid state message: " + messageString);
-      } else {
-        LOGGER.error("Received invalid message: " + messageString);
-        return Optional.empty();
-      }
     }
-  }
-
-  /**
-   * Tests whether the provided JSON string represents a state message.
-   *
-   * @param input a JSON string that represents an {@link AirbyteMessage}.
-   * @return {@code true} if the message is a state message, {@code false} otherwise.
-   */
-  private static boolean isStateMessage(final String input) {
-    final Optional<AirbyteTypeMessage> deserialized = Jsons.tryDeserialize(input, AirbyteTypeMessage.class);
-    return deserialized.filter(airbyteTypeMessage -> airbyteTypeMessage.getType() == Type.STATE).isPresent();
-  }
-
-  /**
-   * Custom class that can be used to parse a JSON message to determine the type of the represented
-   * {@link AirbyteMessage}.
-   */
-  private static class AirbyteTypeMessage {
-
-    @JsonProperty("type")
-    @JsonPropertyDescription("Message type")
-    private AirbyteMessage.Type type;
-
-    @JsonProperty("type")
-    public AirbyteMessage.Type getType() {
-      return type;
-    }
-
-    @JsonProperty("type")
-    public void setType(final AirbyteMessage.Type type) {
-      this.type = type;
-    }
-
+    throw new RuntimeException(String.format("Invalid serialized message: %s", messageString));
```

Comment:
Throwing a RTE because the statement is, some stream is sending bad data that destinations wouldn't be able to handle. Fail fast to not drop data if the data is bad [context](https://github.com/airbytehq/airbyte/pull/27452#discussion_r1234552646)